tests/OutputTest.php

```diff
@@ -0,0 +1,122 @@
+<?php
+namespace Dompdf\Tests;
+
+use CallbackFilterIterator;
+use Dompdf\Tests\TestCase;
+use Dompdf\Tests\OutputTest\Dataset;
+use FilesystemIterator;
+use Iterator;
+use RecursiveDirectoryIterator;
+use RecursiveIteratorIterator;
+use SplFileInfo;
+
+final class OutputTest extends TestCase
+{
+    public const FAILED_OUTPUT_DIRECTORY = __DIR__ . "/../tmp/failed-output-tests";
+
+    /**
+     * @return Iterator<Dataset>
+     */
+    public static function datasets(): Iterator
+    {
+        $flags = FilesystemIterator::KEY_AS_FILENAME
+            | FilesystemIterator::CURRENT_AS_FILEINFO
+            | FilesystemIterator::SKIP_DOTS;
+        $filter = function (SplFileInfo $file) {
+            return $file->getExtension() === "html";
+        };
+        $dir = new RecursiveDirectoryIterator(Dataset::DATASET_DIRECTORY, $flags);
+        $files = new CallbackFilterIterator(new RecursiveIteratorIterator($dir), $filter);
+
+        foreach ($files as $file) {
+            $dataset = new Dataset($file);
+            yield $dataset->name() => $dataset;
+        }
+    }
+
+    public function outputTestProvider(): Iterator
+    {
+        foreach (self::datasets() as $name => $dataset) {
+            yield $name => [$dataset];
+        }
+    }
+
+    /**
+     * @dataProvider outputTestProvider
+     */
+    public function testOutput(Dataset $dataset): void
+    {
+        $document = $dataset->render("gd");
+        $pageCount = $document->getCanvas()->get_page_count();
+        $referenceImages = $dataset->referenceImages();
+
+        if ($pageCount !== count($referenceImages)) {
+            $this->saveFailedOutput($dataset);
+            $this->fail("Output does not match reference rendering (page count does not match)");
+        }
+
+        foreach ($referenceImages as $i => $reference) {
+            $page = $i + 1;
+            $imageData = $document->output(["page" => $page]);
+            $matches = $this->compareImages($reference->getPathname(), $imageData);
+
+            if (!$matches) {
+                $this->saveFailedOutput($dataset);
+                $this->fail("Output does not match reference rendering (difference on page $page)");
+            }
+        }
+
+        $this->addToAssertionCount(1);
+    }
+
+    protected function compareImages(string $referencePath, string $imageData): bool
```

Comment:
This class is `final` so the method should be `private`

Reply:
Also applied this suggestion to the other protected methods.